Mathlib/Data/List/Scan.lean

```diff
@@ -74,17 +88,70 @@ theorem getElem_succ_scanl {i : ℕ} (h : i + 1 < (scanl f b l).length) :
       · simp only [length, Nat.zero_add 1, succ_add_sub_one, hi]; rfl
       · simp only [length_singleton]; omega
 
--- scanr
+/-! ### List.scanr -/
+
+variable {f : α → β → β}
+
 @[simp]
-theorem scanr_nil (f : α → β → β) (b : β) : scanr f b [] = [b] :=
+theorem scanr_nil (b : β) : scanr f b [] = [b] :=
   rfl
 
 @[simp]
-theorem scanr_cons (f : α → β → β) (b : β) (a : α) (l : List α) :
-    scanr f b (a :: l) = foldr f b (a :: l) :: scanr f b l := by
+theorem scanr_ne_nil : scanr f b l ≠ [] := by
+  simp [scanr]
+
+@[simp]
+theorem scanr_cons : scanr f b (a :: l) = foldr f b (a :: l) :: scanr f b l := by
   simp only [scanr, foldr, cons.injEq, and_true]
   induction l generalizing a with
   | nil => rfl
   | cons hd tl ih => simp only [foldr, ih]
 
+@[simp]
+theorem scanr_iff_nil : scanr f b l = [b] ↔ l = [] := by
+  constructor
+  · cases l <;> simp
+  · simp_all
+
+@[simp]
+theorem length_scanr (b : β) (l : List α) : length (scanr f b l) = l.length + 1 := by
+  induction l <;> simp_all
+
+theorem getElem?_scanr_zero : (scanr f b l)[0]? = foldr f b l := by
+  cases l <;> simp
+
+@[simp]
+theorem getElem_scanr_zero : (scanr f b l)[0] = foldr f b l := by
+  cases l <;> simp
+
+@[simp]
+theorem head_scanr : (scanr f b l).head scanr_ne_nil = foldr f b l := by
+  cases l <;> simp
+
+theorem tail_scanr (h : 0 < l.length) : (scanr f b l).tail = scanr f b l.tail := by
+  induction l <;> simp_all
+
+theorem drop_scanr {i : ℕ} (h : i < l.length + 1) :
+    (scanr f b l).drop i = scanr f b (l.drop i) := by
+  induction i with
+  | zero => simp
+  | succ i ih =>
+      rw [← drop_drop]
+      simp [ih (by omega), tail_scanr (l := l.drop i) (by rw [length_drop]; omega)]
+
+theorem getElem_scanr {i : ℕ} (h : i < l.length + 1) :
+    (scanr f b l)[i]'(by simp [h]) = foldr f b (l.drop i) := by
+  induction l generalizing i with
+  | nil => simp [h]
+  | cons head tail ih =>
+      by_cases h' : i = 0
```

Comment:
A common pattern we use is `obtain rfl | h' := eq_or_ne i 0` for this, which might be nice to use here too